providers/common/ciphers/aes.c

```diff
@@ -255,69 +248,82 @@ static int aes_cipher(void *vctx,
     return 1;
 }
 
-#define IMPLEMENT_new_params(lcmode, UCMODE) \
-    static OSSL_OP_cipher_get_params_fn aes_##lcmode##_get_params; \
-    static int aes_##lcmode##_get_params(OSSL_PARAM params[]) \
+#define IMPLEMENT_cipher(lcmode, UCMODE, flags, klen, blklen, ivlen) \
+    static OSSL_OP_cipher_get_params_fn aes_##klen##_##lcmode##_get_params; \
+    static int aes_##klen##_##lcmode##_get_params(OSSL_PARAM params[]) \
     { \
         OSSL_PARAM *p; \
-    \
-        p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_MODE); \
-        if (p != NULL && !OSSL_PARAM_set_int(p, EVP_CIPH_##UCMODE##_MODE)) \
-            return 0; \
+                                                                \
+        p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_MODE);          \
+        if (p != NULL) {                                                \
+            if (!OSSL_PARAM_set_int(p, EVP_CIPH_##UCMODE##_MODE))           \
+                return 0;                                               \
+        }                                                           \
+        p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_FLAGS); \
+        if (p != NULL) {                                                \
+            if (!OSSL_PARAM_set_ulong(p, (flags)))                          \
+                return 0;                                               \
+        }                                                           \
+        p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_KEYLEN);        \
+        if (p != NULL) {                                                \
+            if (!OSSL_PARAM_set_int(p, (klen) / 8))                         \
+                return 0;                                               \
+        }                                                           \
+        p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_BLOCK_SIZE);    \
+        if (p != NULL) {                                                \
+            if (!OSSL_PARAM_set_int(p, (blklen) / 8))                   \
+                return 0;                                               \
+        }                                                               \
+        p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_IVLEN);         \
+        if (p != NULL) {                                                \
+            if (!OSSL_PARAM_set_int(p, (ivlen) / 8))                    \
+                return 0;                                               \
+        }                                                               \
     \
         return 1; \
-    }
-
-#define IMPLEMENT_new_ctx(lcmode, UCMODE, len) \
-    static OSSL_OP_cipher_newctx_fn aes_##len##_##lcmode##_newctx; \
-    static void *aes_##len##_##lcmode##_newctx(void *provctx) \
+    } \
+    static OSSL_OP_cipher_newctx_fn aes_##klen##_##lcmode##_newctx; \
+    static void *aes_##klen##_##lcmode##_newctx(void *provctx) \
```

Comment:
maybe we should be using 'bits' instead of 'len' for the names.

Reply:
Sure, I can do that.

Reply:
Done